homeassistant/components/azure_devops/__init__.py

```diff
@@ -1,80 +1,104 @@
 """Support for Azure DevOps."""
 from __future__ import annotations
 
+from datetime import timedelta
 import logging
+from typing import Final
 
+from aioazuredevops.builds import DevOpsBuild
 from aioazuredevops.client import DevOpsClient
+from aioazuredevops.core import DevOpsProject
 import aiohttp
 
 from homeassistant.config_entries import ConfigEntry
 from homeassistant.core import HomeAssistant
-from homeassistant.exceptions import ConfigEntryAuthFailed, ConfigEntryNotReady
-from homeassistant.helpers.entity import DeviceInfo, Entity
+from homeassistant.exceptions import ConfigEntryAuthFailed
+from homeassistant.helpers.entity import DeviceInfo
+from homeassistant.helpers.update_coordinator import (
+    CoordinatorEntity,
+    DataUpdateCoordinator,
+    UpdateFailed,
+)
 
-from .const import CONF_ORG, CONF_PAT, CONF_PROJECT, DATA_AZURE_DEVOPS_CLIENT, DOMAIN
+from .const import CONF_ORG, CONF_PAT, CONF_PROJECT, DOMAIN
 
 _LOGGER = logging.getLogger(__name__)
 
 PLATFORMS = ["sensor"]
 
+BUILDS_QUERY: Final = "?queryOrder=queueTimeDescending&maxBuildsPerDefinition=1"
+
 
 async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     """Set up Azure DevOps from a config entry."""
-    client = DevOpsClient()
-
-    try:
-        if entry.data[CONF_PAT] is not None:
-            await client.authorize(entry.data[CONF_PAT], entry.data[CONF_ORG])
-            if not client.authorized:
-                raise ConfigEntryAuthFailed(
-                    "Could not authorize with Azure DevOps. You may need to update your token"
-                )
-        await client.get_project(entry.data[CONF_ORG], entry.data[CONF_PROJECT])
-    except aiohttp.ClientError as exception:
-        _LOGGER.warning(exception)
-        raise ConfigEntryNotReady from exception
-
-    instance_key = f"{DOMAIN}_{entry.data[CONF_ORG]}_{entry.data[CONF_PROJECT]}"
-    hass.data.setdefault(instance_key, {})[DATA_AZURE_DEVOPS_CLIENT] = client
+    client: DevOpsClient = DevOpsClient()
+
+    async def async_update_data() -> tuple[
+        DevOpsClient, DevOpsProject, list[DevOpsBuild]
+    ]:
+        """Fetch data from Azure DevOps."""
+
+        try:
+            if entry.data[CONF_PAT] is not None:
+                await client.authorize(entry.data[CONF_PAT], entry.data[CONF_ORG])
+                if not client.authorized:
+                    raise ConfigEntryAuthFailed(
+                        "Could not authorize with Azure DevOps. You may need to update your token"
```

Comment:
```suggestion
                        "Could not authorize with Azure DevOps. You need to update your token"
```

If it fails, it's not a question, they have to do that :)